### PR TITLE
Add upstream CI to pyarn repo

### PR DESCRIPTION
**Pull request by fepas**

### CODE REVIEW

**Comment by fepas** [comment]
https://github.com/fepas/pyarn/actions The tests/linters passing on my fork

File: pyarn/lexer.py [hunk 1]
```diff
@@ -15,36 +15,36 @@
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
 """
 tokens = (
-    'STRING',
```
Comment (MartinBasti) [inline pyarn/lexer.py:18]: Why `"` are wrong?

**Comment by MartinBasti** [comment]
`Fix tox problems` this commit is quite big, it's hard to understand which linter and why it needs to be fixed. Would be possible to split it into multiple commits, commit per type of fix?

File: pyarn/lexer.py (hunk 1 above)
Comment (taylormadore) [reply pyarn/lexer.py:18]: Black formatting changes maybe? If so, I wonder if we could do all of that in a separate commit

Comment (fepas) [reply pyarn/lexer.py:18]: I think we should ask flake8/black, I just followed the suggestions :sweat_smile:

Comment (MartinBasti) [reply pyarn/lexer.py:18]: ah, it's black

**Comment by fepas** [comment]
@MartinBasti @taylormadore I rewrote the commits in a more cohesive way, thanks :D

**Comment by MartinBasti** [comment]
Diff from black is quite overwhelming, and because there are changes in string and regexp, reviewers must be cautions, do we really need it? It adds indirection to git history.

File: requirements-dev.txt [hunk 2]
```diff
@@ -1,3 +1,11 @@
...
 pytest
 build
+tox
+isort
+mypy
+black==22.6.0
+click==8.0.3
+bandit
+pytest
+pytest-cov
```
Comment (chmeliik) [inline requirements-dev.txt]: Are these intended to be here? Tox will install the specified `deps` to separate virtual envs, they don't need to be duplicated here

(alternatively, they can be kept here and don't need to be duplicated in tox.ini)

File: .coveragerc [hunk 3]
```diff
@@ -0,0 +1,8 @@
+[report]
+skip_covered = True
+show_missing = True
+fail_under = 75
```
Comment (chmeliik) [inline .coveragerc]: I don't think it needs to be that low, maybe 95?

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+[testenv]
+deps =
+    -rrequirements.txt
+    -rrequirements-dev.txt
+whitelist_externals =
+    make
+    mkdir
+    rm
+passenv = TOX_ENV_DIR
```
Comment (chmeliik) [inline tox.ini]: Do we need any of this?

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+usedevelop = true
+
+commands =
+    pytest \
+        --cov-config .coveragerc --cov=pyarn --cov-report term \
+        --cov-report xml --cov-report html
```
Comment (chmeliik) [inline tox.ini]: Are xml and html useful?

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+[testenv:black]
+description = black checks [Mandatory]
+skip_install = true
+deps =
+    # Pin the version of black and click to avoid a newer version causing tox to fail
+    black==22.6.0
+    click==8.0.3
```
Comment (chmeliik) [inline tox.ini]: Are the newer versions causing tox to fail?

File: tox.ini [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+[flake8]
+show-source = True
+max-line-length = 100
+exclude = venv,.git,.tox,.env,pyarn/parsetab.py,pyarn/parser.out
+ignore = D102,D107,D101,D103,D105,D400,E226,D300,D104
+per-file-ignores =
+    # Ignore missing docstrings in the tests, migrations and missing/multiple whitespace
+    tests/*:D101,D102,D103,E226,E241,E231
+    ./pyarn/parser.py: D400,D403,D205,D209,D208,D103
+;     /tests/test_lexer.py:
```
Comment (chmeliik) [inline tox.ini]: It would be good to explain what we are ignoring and justify why

File: tox.ini [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+deps =
+    bandit
+commands =
+    bandit -r pyarn
+
+[testenv:mypy]
+skip_install = true
+commands =
+    mypy -p pyarn --install-types --non-interactive \
+    --exclude pyarn/parsetab.py --ignore-missing-imports
```
Comment (chmeliik) [inline tox.ini]: Do we have to `--ignore-missing-imports`?

File: tox.ini [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+skip_install = true
+deps =
+    flake8==3.9.2
+    flake8-docstrings==1.6.0
+commands =
+    flake8
+
+[flake8]
+show-source = True
+max-line-length = 100
```
Comment (chmeliik) [inline tox.ini]: We should configure black and isort to use line length 100 as well, or use black's default 88 for all three

File: pyarn/parser.py [hunk 10]
```diff
@@ -56,21 +56,21 @@ def p_block_comment(p):
 
 def p_title(p):
     """title : STRING COLON INDENT
-             | list COLON INDENT"""
+    | list COLON INDENT"""
```
Comment (chmeliik) [inline pyarn/parser.py]: We should preserve the formatting here. Maybe like this?

```python
"""
title : STRING COLON INDENT
      | list COLON INDENT
"""
```

**Comment by chmeliik** [comment]
> Diff from black is quite overwhelming, and because there are changes in string and regexp, reviewers must be cautions, do we really need it? It adds indirection to git history.

I don't have a problem with it apart from the misaligned docstrings (https://github.com/containerbuildsystem/pyarn/pull/16#discussion_r1021258058). I don't think that much caution is needed, black doesn't make semantic changes (except for [special cases](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#ast-before-and-after-formatting))

File: tox.ini (hunk 9 above)
Comment (chmeliik) [reply tox.ini]: And we might want to use `profile = "black"` for isort https://pycqa.github.io/isort/docs/configuration/black_compatibility.html

File: tox.ini [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+[gh-actions]
+python =
+    3.9: python3.9
+    3.10: python3.10
+
+[testenv]
+deps =
+    -rrequirements.txt
+    -rrequirements-dev.txt
+whitelist_externals =
```
Comment (ejegrova) [inline tox.ini]: whitelist_externals is deprecated in favor of allowlist_externals: https://tox.wiki/en/latest/config.html#conf-allowlist_externals

File: requirements-dev.txt (hunk 2 above)
Comment (fepas) [reply requirements-dev.txt]: This is just there so the developer can test outside of tox, using venv (and apply the formatting commands).

File: tox.ini (hunk 4 above)
Comment (fepas) [reply tox.ini]: I don't think so, I'll remove it

File: tox.ini (hunk 6 above)
Comment (fepas) [reply tox.ini]: I don't know, I just kept the version that is in Cachi2. Sounds like a safe approach to me.

File: tox.ini (hunk 5 above)
Comment (fepas) [reply tox.ini]: I don't think so, I'll remove both of them.

File: requirements-dev.txt (hunk 2 above)
Comment (MartinBasti) [reply requirements-dev.txt]: but tox is the easy way developers to test it

File: tox.ini (hunk 6 above)
Comment (MartinBasti) [reply tox.ini]: is `click` even used in this repo?

File: tox.ini (hunk 8 above)
Comment (MartinBasti) [reply tox.ini]: we have to test, in OSBS adn cachito, not all deps/imports had definitions for mypy

Comment (fepas) [reply tox.ini]: Yes, instead we get: 

> mypy run-test: commands[0] | mypy -p pyarn --install-types --non-interactive --exclude pyarn/parsetab.py
pyarn/lockfile.py:22: error: Skipping analyzing "ply": module is installed, but missing library stubs or py.typed marker  [import]
pyarn/lockfile.py:22: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 5 source files)

File: tox.ini (hunk 11 above)
Comment (MartinBasti) [reply tox.ini]: do we need externals in pyarn?

File: tox.ini (hunk 8 above)
Comment (chmeliik) [reply tox.ini]: Ah, fair enough. But now I'm wondering if `mypy` really has any value here :sweat_smile: 
* the only external dependency, `ply`, does not have type annotations
* there's not a single type annotation in the pyarn code
* and most importantly, mypy *does not check untyped methods by default*
  * https://mypy.readthedocs.io/en/stable/config_file.html#confval-check_untyped_defs

So in short, mypy checks absolutely nothing here :smile: 

We could at least enable `check_untyped_defs` and get some minimal value out of it I guess

Comment (MartinBasti) [reply tox.ini]: IIRC mypy is able to do some basic checks with even without annotations. Because it found some bugs inside OSBS w/o annotations

File: requirements-dev.txt (hunk 2 above)
Comment (chmeliik) [reply requirements-dev.txt]: Maybe we can have the best of both worlds by maintaining the tox deps _only_ in this file? And remove all the duplicated `deps` sections from tox.ini?

File: tox.ini (hunk 8 above)
Comment (chmeliik) [reply tox.ini]: With `--check-untyped-defs` yes, without it I don't think mypy checks anything
```python
def obviously_wrong():
    a = 1 + "two"


def obviously_wrong2() -> None:
    a = 1 + "two"
```
```shell
$ mypy x.py 
x.py:6: error: Unsupported operand types for + ("int" and "str")
Found 1 error in 1 file (checked 1 source file)

$ mypy x.py --check-untyped-defs
x.py:2: error: Unsupported operand types for + ("int" and "str")
x.py:6: error: Unsupported operand types for + ("int" and "str")
Found 2 errors in 1 file (checked 1 source file)
```

Comment (fepas) [reply tox.ini]: Since mypy doesn't seem to add anything, I'll remove it.

File: tox.ini [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    flake8-docstrings==1.6.0
+commands =
+    flake8
+
+[flake8]
+show-source = True
+max-line-length = 100
+exclude = venv,.git,.tox,.env,pyarn/parsetab.py,pyarn/parser.out
+ignore = D101,D102,D103,D104,D105,D107,D300,D400,E226
+; Missing Docstrings
```
Comment (brunoapimentel) [inline tox.ini]: I think the comments explaining what the ignore codes mean should be in the line before the actual `ignore` definition:

 ```
; Ignored  checks
; D101	Missing docstring in public class
; D102	Missing docstring in public method
; D103	Missing docstring in public function
; D104	Missing docstring in public package
; D105	Missing docstring in magic method
; D107	Missing docstring in __init__
; D300	Use “”“triple double quotes”“”
; D400	First line should end with a period
; E226	Missing whitespace around arithmetic operator
ignore = D101,D102,D103,D104,D105,D107,D300,D400,E226
```

File: tox.ini [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+; Quotes Issues
+; D300	Use “”“triple double quotes”“”
+
+; Docstring Content Issues
+; D400	First line should end with a period
+
+; Whitespace Issues
+; E226	Missing whitespace around arithmetic operator
+
+per-file-ignores =
```
Comment (brunoapimentel) [inline tox.ini]: Here I would also change the comments to be in the lines before the actual declaration, and follow the indent level.
```
per-file-ignores =
    ; Whitespace Issues
    ; E231	Missing whitespace after ',', ';', or ':'
    ; E241	Multiple spaces after ','
    tests/*:E231,E241
    
    ; D205	1 blank line required between summary line and description
    ; D208	Docstring is over-indented
    ; D209	Multi-line docstring closing quotes should be on a separate line
    ; D403	First word of the first line should be properly capitalized
    ./pyarn/parser.py:D205,D208,D209,D403
```

**Comment by MartinBasti** [comment]
Those new commits should be squashed to original ones. I don't see reason why the same PR should add and revert at the same time.

File: tox.ini [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+; E231	Missing whitespace after ',', ';', or ':'
+; E241	Multiple spaces after ','
+    ./pyarn/parser.py:D205,D208,D209,D403
+; Whitespace Issues
+; D205	1 blank line required between summary line and description
+; D208	Docstring is over-indented
+; D209	Multi-line docstring closing quotes should be on a separate line
+
+; Docstring Content Issues
+; D403	First word of the first line should be properly capitalized
```
Comment (chmeliik) [inline tox.ini]: I have a different suggestion for the ignores: how about remove all of them? :smile: 

* all the whitespace related issues should be solved by `black`
* instead of ignoring every docstring related issue, how about we don't use flake8-docstrings at all?

File: tox.ini [hunk 15]
```diff
@@ -10,30 +10,24 @@ python =
...
-    # Pin the version of black and click to avoid a newer version causing tox to fail
+    # Pin the version of black to avoid a newer version causing tox to fail
     black==22.6.0
-    click==8.0.3
 commands =
     black --check --diff pyarn tests
 
 [testenv:isort]
 skip_install = true
+profile = "black"
```
Comment (chmeliik) [inline tox.ini]: Careful, `[testenv:isort]` is Tox configuration, not isort configuration. The configuration for isort would go in `[isort]` or a separate project file (like setup.cfg)

https://pycqa.github.io/isort/docs/configuration/config_files.html

File: tox.ini (hunk 9 above)
Comment (chmeliik) [reply tox.ini]: I might be blind, but I don't see line length configured anywhere :smile:

**Comment by fepas** [comment]
@chmeliik @ejegrova @MartinBasti @brunoapimentel @taylormadore 

I believe that all the problems pointed out so far have been resolved, take a look when you can!

**Review by ejegrova — APPROVED** [review]
LGTM

File: setup.cfg [hunk 16]
```diff
@@ -32,4 +32,12 @@ install_requires =
...
-exclude = pyarn/parsetab.py
+exclude = venv,.tox,pyarn/parsetab.py,pyarn/parser.out
+show-source = True
+
+[isort]
+profile = black
+line_length = 100
+
+[black]
+line_length = 100
```
Comment (chmeliik) [inline setup.cfg]: Does black really read setup.cfg? I feel there should have been formatting changes after bumping the line length, but there were none

File: setup.cfg [hunk 17]
```diff
@@ -32,4 +32,12 @@ install_requires =
 [flake8]
 max-line-length = 100
 doctests = True
-exclude = pyarn/parsetab.py
+exclude = venv,.tox,pyarn/parsetab.py,pyarn/parser.out
+show-source = True
```
Comment (chmeliik) [inline setup.cfg:36]: +1 for keeping flake8 config only in setup.cfg :+1:

**Review by chmeliik — COMMENTED** [review]
Just minor issues left :tada: 

One more thing: I remembered the Makefile has a `check` command for running tests, we should update that to run `tox`

File: setup.cfg (hunk 16 above)
Comment (taylormadore) [reply setup.cfg]: Looks like black uses pyproject.toml: https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html#black-compatible-configurations

Comment (fepas) [reply setup.cfg]: The default for black is 88, put it there just to be sure :D

Comment (taylormadore) [reply setup.cfg]: Aren't we using line length 100 everywhere else though? When I add
```
[tool.black]
line-length = 100
```
to pyproject.toml, I see black making a few formatting changes

**Comment by taylormadore** [comment]
Other than that last minor question, this is looking pretty good to me :tada:

**Review by chmeliik — APPROVED** [review]
LGTM :+1:
